rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,117 @@
+    └── generic/ # Everything that is related to generic kernels
+        └── sycl/ # generic SYCL kernels and related abstractions
+
+```
+
+### Affected Basic Abstractions
+
+The new schema will require moving a lot of parts of the library around
+and while most of the changes are probably just an implementation detail
+there are a few major changes that have to be described in this RFC.
+
+#### Engine
+
+Engine is the most heavily affected abstraction by the changes.
+There is currently no `gpu_engine_t` class as `compute_engine_t` is used as
+an Intel centric alternative of it. In order to have a common basic
+abstraction at the GPU level the `gpu_engine_t` class will be introduced.
+
+The current inheritance chains for SYCL and OpenCL GPU engines are
+the following:
+* SYCL Intel and generic: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_engine_t`
+* SYCL NVIDIA: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_cuda_engine_t`
+* SYCL AMD: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_hip_engine_t`
+* OpenCL (only Intel): `engine_t` -> `compute_engine_t` -> `ocl_gpu_engine_t`
+
+With the new schema they will be changed to:
+* SYCL Intel and generic: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_intel_engine_t`
+* SYCL NVIDIA: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_nvidia_engine_t`
+* SYCL AMD: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_amd_engine_t`
+* OpenCL: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `ocl_gpu_engine_t`
+    * There is currently no plan to extend OpenCL support to other vendors
+    therefore the name of the final class (`ocl_gpu_engine_t`) doesn't
+    include the vendor's name.
+
+There doesn't seem to be much sense in keeping `compute_engine_t` as
+`gpu_intel_engine_t` can take over its responsibilities.
+
+#### Stream
+
+The current inheritance chains for SYCL and OpenCL GPU streams are the following:
+* SYCL Intel and genetic: `stream_t` -> `compute_stream_t` -> `sycl_stream_t`
+* SYCL NVIDIA: `stream_t` -> `compute_stream_t` -> `sycl_stream_t` -> `sycl_cuda_stream_t`
+* SYCL AMD: `stream_t` -> `compute_stream_t` -> `sycl_stream_t` -> `sycl_hip_stream_t`
+* OpenCL: `stream_t` -> `compute_stream_t` -> `ocl_stream_t`
+
+With the new schema they will be changed to:
+The current inheritance chains for SYCL and OpenCL GPU streams are the following:
+* SYCL Intel and genetic: `stream_t` -> `gpu_stream_t` -> (`gpu_intel_stream_t` -> `compute_stream_t`) -> `sycl_intel_stream_t`
+* SYCL NVIDIA: `stream_t` -> `gpu_stream_t` -> (`gpu_intel_stream_t` -> `compute_stream_t`) -> `sycl_nvidia_stream_t`
+* SYCL AMD: `stream_t` -> `gpu_stream_t` -> (`gpu_intel_stream_t` -> `compute_stream_t`) -> `sycl_amd_stream_t`
```

Comment:
Does it make sense for AMD and NVidia streams to inherit from streamwith Intel in its name - `gpu_intel_stream_t`?

Reply:
Fixed it.